### ISSUE TITLE
connection.query - no return and no error

### ISSUE THREAD

**Issue by DaveyD** [nuintun/node-adodb#55]
Hi, 
I've been using this a lot and it works great!
However, I am running a query now and I am not getting any output, no records and no error message.
It seems that it is not getting to the "then()" or the "catch"

I am using:
Node: 8.11.1
node-adodb: 4.0.9

Again, I have other queries that work perfect, but this one is not producing any output.

Here is the code: (the sql was copied directly out of access where it works perfect)
[P.S. I formatted the sql for this post for easier readability - in actual code eache statement is on 1 line as in access]
```` 

var ADODB = require('node-adodb');
ADODB.debug = true;

var accessDB = "D:\\PATH\\TO\\MY\\ACCESS_DB.accdb"
var conn = ADODB.open('Provider=Microsoft.ACE.OLEDB.12.0;Data Source='+accessDB+';');
var sql = `
    SELECT
        tbl_Books.SKU,
        lst_Books.BookName AS BookID,
        lst_Sizes.SizeName AS [Size] 
    FROM
        lst_Sizes 
        INNER JOIN
            (
                lst_Books 
                INNER JOIN
                    tbl_Books 
                    ON lst_Books.ID = tbl_Books.Book
            )
            ON lst_Sizes.ID = tbl_Books.Size 
    ORDER BY
        lst_Books.BookName,
        lst_Sizes.SizeName;
`

conn
   .query(sql)
   .then(data => {
      console.log(JSON.stringify(data, null, 2));
      // cb_updateData(data);
   })
   .catch(error => {
      console.error(error);
   });
`````

Thanks,
David

**Comment by DaveyD** [comment]
I just did more testing and it seems that it is the INNER JOIN statement that is causing the problem
In the above example there are 2 INNER JOINs - table 2 and 3 are joined to 1
If I remove one of those tables from the JOIN then the query works perfect.
E.g. 
````
  SELECT tbl_Books.SKU, lst_Books.BookName AS BookID
    FROM lst_Books INNER JOIN tbl_Books ON lst_Books.ID = tbl_Books.Book
    ORDER BY lst_Books.BookName;
````
The above sql works perfectly

Any help would be greatly appreciated
Thanks,
David

**Comment by nuintun** [comment]
@DaveyD 
![_20180502113333](https://user-images.githubusercontent.com/1730277/39504218-aafad4ce-4dfc-11e8-834f-3da8aa86e026.png)

The sql is invalid!

Try this `query.js`
```
var connection = new ActiveXObject('ADODB.Connection');
var recordset = new ActiveXObject('ADODB.Recordset');

/**
 * @function free
 * @description Free connection
 * @param {Object} connection
 */
function free(connection) {
  connection.State && connection.Close();
}

/**
 * @function isDate
 * @param {number} value
 * @returns {boolean}
 * @see https://docs.microsoft.com/en-us/sql/ado/reference/ado-api/datatypeenum
 */
function isDate(value) {
  return value === 7 || value === 64 || value === 133 || value === 134 || value === 135;
}

/**
 * @function isBinary
 * @param {number} value
 * @returns {boolean}
 * @see https://docs.microsoft.com/en-us/sql/ado/reference/ado-api/datatypeenum
 */
function isBinary(value) {
  return value === 128 || value === 204 || value === 205;
}

/**
 * @function readBinary
 * @param {Binary} binary
 * @returns {string}
 */
function readBinary(binary) {
  var stream = new ActiveXObject('ADODB.Stream');

  stream.Type = 1;

  stream.Open();

  stream.Position = 0;

  stream.Write(binary);

  stream.Position = 0;
  stream.Type = 2;

  var text = stream.ReadText();

  free(stream);

  return text;
}

/**
 * @function fillRecords
 * @param {Recordset} recordset
 * @returns {Array}
 */
function fillRecords(recordset) {
  var records = [];
  var fields = recordset.Fields;

  // Not empty
  if (!recordset.BOF || !recordset.EOF) {
    var i, record;
    var field, type, value;
    var count = fields.Count;

    recordset.MoveFirst();

    while (!recordset.EOF) {
      record = {};

      for (i = 0; i < count; i++) {
        field = fields.Item(i);
        type = field.Type;
        value = field.Value;

        // ADO has given us a UTC date but JScript assumes it's a local timezone date
        // Thanks https://github.com/Antony74
        if (isDate(type)) {
          value = new Date(value);
        } else if (isBinary(type)) {
          value = readBinary(value);
        }

        record[field.Name] = value;
      }

      WScript.StdOut.Write(field.value);

      records.push(record);
      recordset.MoveNext();
    }
  }

  // Return records
  return records;
}

// Set CursorLocation
connection.CursorLocation = 3;

// Open
connection.Open('Provider=Microsoft.ACE.OLEDB.12.0;Data Source=node-adodb.accdb;');

// Query
recordset.Open(
  'SELECT tbl_Books.Book, lst_Sizes.SizeName FROM lst_Sizes INNER JOIN tbl_Books ON lst_Sizes.ID = tbl_Books.Size;',
  connection,
  0,
  1
);

// Write data
fillRecords(recordset);

// Close database
free(recordset);
free(connection);
```

Run `cscript query.js`

This error no description with catch, so stderr not work - -!

**Comment by DaveyD** [comment]
@nuintun,
Thanks for the repsonse!
I just tried your code and got the same result as you `(null): unspecified error`
However, I dont understand why the sql is invalid - I copied it right out of Access!

You wrote:
> I will fix this soon.

Do you mean:
1. You will fix it so the sql will work, or
1. You will fix the error reporting to show this error?

If you mean number 2 (just the error), how then can I get this sql to work correctly?


Thanks again!
David

**Comment by nuintun** [comment]
See [INNER JOIN Operation](https://support.office.com/en-us/article/inner-join-operation-b9e73ab6-884a-403e-9f22-cb502feae36a?omkt=en-US&ui=en-US&rs=en-US&ad=US).

> You can join any two numeric fields of like types. For example, you can join on AutoNumber and Long fields because they are like types. However, you cannot join Single and Double types of fields.

You can try this sql, it's ok!
```sql
SELECT tbl_Books.Book, lst_Sizes.SizeName FROM lst_Sizes INNER JOIN tbl_Books ON lst_Sizes.SizeName = tbl_Books.SKU;
```

**Comment by DaveyD** [comment]
@nuintun , thanks for response
The query that you supplied does work and returns an empty array - that's because the fields you connected dont match.

After a little more testing, i realized that the problem is that the table field is a lookup field!
Lookup fields are numbers so technically they should work, but seemingly they dont
I made a new "number" field in the table containing the value of that lookup field, and querying on that field does work perfectly.
So, in short, the problem is using a lookup field in the INNER JOIN
Is this supposed to be this way or is this some type of bug?

Please let me know
Thanks,
David

**Comment by DaveyD** [comment]
@nuintun ,
I just did some more testing and I am realizing that it is not only a lookup field problem
I changed all my lookup fields to be regular numbers and it is still not working

To show the very weird issue, I made another sample DB
This db has 3 tables:
* lst_Sizes (contains 2 fields: numeric ID and shortText Size)
* lst_Books (contains 2 fields: numeric ID and shortText Size)
* tbl_Books (contains 3 fields: numeric ID, numeric Book, numeric Size)

All fields are formatted the exact same way and therefore the JOINS should work properly
**However:**
When I query INNER JOIN tbl_books on lst_Books, it returns results as expected
But when I run the same exact query on lst_Sizes, I get the same `(null) unspecified error`

I tested this using your query.js script that you posted above
Here are the 2 queries that I tried:

**1: Book** (Works perfectly)
```sql
'SELECT tbl_Books.SKU, lst_Books.BookName FROM lst_Books INNER JOIN tbl_Books ON lst_Books.ID = tbl_Books.Book;'
```

**2: Size** (returns null)
```sql
'SELECT tbl_Books.SKU, lst_Sizes.SizeName FROM lst_Sizes INNER JOIN tbl_Books ON lst_Sizes.ID = tbl_Books.Size;'
```

Please let me know
Thanks,
David


P.S. Reminder, both queries work fine within access
There is a query in the access file for each one - you can see there

**Comment by nuintun** [comment]
Use:

```sql
SELECT tbl_Books.SKU, lst_Sizes.SizeName FROM lst_Sizes INNER JOIN tbl_Books ON lst_Sizes.ID = tbl_Books.[Size];
```

See:

[Access 2007 reserved words and symbols](https://support.office.com/en-us/article/Access-2007-reserved-words-and-symbols-e33eb3a9-8baa-4335-9f57-da237c63eabe?CorrelationId=0114329f-35f7-4be7-b405-124d0b5fec9a&ui=en-US&rs=en-US&ad=US&ocmsassetID=HA010030643)

**Comment by DaveyD** [comment]
@nuintun , GREAT!
That was the problem!!
Now all my past sqls are working!!

Is there a way that in the future we can get some type of hint to what the error is?
(Access did not have an issue with it... so it's hard to figure out)

Thanks,
David

P.S. I didnt close the issue just in case it should be a reminder to implement the fix
Bur, you can close it when you're ready (or I can, just let me know)

**Comment by nuintun** [comment]
Fixed, add a message when error!
